topmodel/utils.jl

```diff
@@ -0,0 +1,96 @@
+"""
+    regrid_and_compute_statistic(data, (lon, lat), resolution, statistics)
+
+Regrids `data` to the resolution specified by computing statistics of interest
+over all data points in the simulation grid boxes of size resolution.
+
+This function takes `data` (size (nlon, nlat)) defined at the points specified by the arrays of `lon` (length nlon) and `lat` (length nlat), 
+as well as a desired `resolution` in degrees. The resolution should
+correspond to grid cells larger than 1kmx1km (the resolution of `data`).
+
+For each coarse grid cell, the statistics are computed over the data in that cell
+, ignoring points in lakes or oceans.
+
+They must be a function of the form x::AbstractArray - > f(x)::Float.
+"""
+function regrid_and_compute_statistics(data, (lon, lat), resolution, statistics)
+    (lat_min, lat_max) = extrema(lat)
+    (lon_min, lon_max) = extrema(lon)
+
+    lat_count = Int(ceil((lat_max - lat_min) / resolution)) + 1
+    lon_count = Int(ceil((lon_max - lon_min) / resolution)) + 1
+
+    n_stats = length(statistics)
+    outdata = zeros(Float32, lon_count, lat_count, n_stats)
+    lat_mask = BitArray(zeros(size(lat)));
+    lon_mask = BitArray(zeros(size(lon)));
+    for lat_id in 1:1:lat_count
+        for lon_id in 1:1:lon_count
+            @show lat_id/lat_count
+            lat_mask .=
+                (lat .>= lat_min + resolution * (lat_id - 1)) .&
+                (lat .< lat_min + resolution * lat_id)
+            lon_mask .=
+                (lon .>= lon_min + resolution * (lon_id - 1)) .&
+                (lon .< lon_min + resolution * lon_id)
+            x = data[lon_mask, lat_mask];
+            # If `x` is Missing, we are over the ocean, and not over the land.
+            # Here we make a land mask by checking where x is *not* Missing.
+            x_land_mask = .!ismissing.(x)
+            if sum(x_land_mask) / prod(size(x)) > 0.5 # count as land
+                for (i, stat) in enumerate(statistics)
+                    outdata[lon_id, lat_id, i] = stat(x[x_land_mask])
+                end
+            else
+                outdata[lon_id, lat_id, :] .= 0f0 # all set to zero
+            end
+
+        end
+    end
+    return outdata,
+    range(stop = lat_max, step = resolution, length = lat_count),
+    range(stop = lon_max, step = resolution, length = lon_count)
+end
```

Comment:
```suggestion
function regrid_and_compute_statistics(data, (lon, lat), resolution, statistics)
    lat_min, lat_max = extrema(lat)
    lon_min, lon_max = extrema(lon)

    lat_grid = range(lon_min, step=resolution, length=ceil(Int, (lon_max - lon_min) / resolution) + 1)
    lon_grid = range(lat_min, step=resolution, length=ceil(Int, (lat_max - lat_min) / resolution) + 1)

    n_stats = length(statistics)
    outdata = zeros(Float32, length(lon_grid), length(lat_grid), n_stats)

    for (lat_id, lat_val) in enumerate(lat_grid)
        for (lon_id, lon_val) in enumerate(lon_grid)
            lat_mask = (lat .>= lat_val) .& (lat .< lat_val + resolution)
            lon_mask = (lon .>= lon_val) .& (lon .< lon_val + resolution)

            x = data[lon_mask, lat_mask]
            x_land_mask = .!ismissing.(x)

            if sum(x_land_mask) / length(x) > 0.5
                for (i, stat) in enumerate(statistics)
                    outdata[lon_id, lat_id, i] = stat(x[x_land_mask])
                end
            end # Implicitly sets to zero if condition not met
        end
    end

    return outdata, lat_grid, lon_grid
end
```

Reply:
This function looks a little sketchy, there's probably better ways to write it.

Reply:
Idk, it works, and it isnt slow. We need to go through each grid box and find all elements within the grid box, and then compute statistics on it, and I wrote it generally so that it will work for an arbitrary number of statistics.